Silostaking/s2 blocks

m4ksio opened the pull request:
Replaces https://github.com/SiloMEV/sei-chain/pull/10/files

Adds very rudimentary MEV implementation

@tychoish

Commit:
MEV

Commit:
Rough draft of MEV Module

Commit:
Step 2

Commit:
Basic bundle querying and submission via cli

Commit:
temp test

Commit:
Tests and build working; and basic query and tx commands.

Commit:
readme for module

Commit:
Architecture for bundle submission

Commit:
Bundle Processing - go test passing & make install passing

Commit:
Updated Go import for silomev tendermint

Commit:
Reverting all ante handler changes.

Commit:
Purge priority change

Commit:
Update Prepare Proposal to query and includes bundles

Commit:
Additional tests: first commit

Commit:
V1 with integration test

Commit:
readme edit

Commit:
Updates for ephermeral storage

m4ksio committed:
Refactor MEV transaction data and parsing logic.

Replace `string` with `bytes` for transaction data to enhance efficiency. Updated related parsing functions and protobuf definitions to reflect the new type. Removed an unused utility function and replaced its usage with an Ethereum library function.

m4ksio committed:
test fixes

m4ksio committed:
Use GO_MOD_CACHE for Go module cache path

Replaced GO_PKG_PATH with GO_MOD_CACHE to better leverage the go command's environment settings for module caching. This ensures consistency with Go's tooling and simplifies configuration by relying on the GOMODCACHE environment variable. Updated relevant paths in Makefile and docker-compose.yml.

m4ksio committed:
Added grpc types registration, so they are actually used. Updated integration test to follow the prescribed schema
Adjusted related tests and modules accordingly.

m4ksio committed:
Shape basic MEV integration test
Remove sender validation

m4ksio committed:
Skeleton work for MEV off-chain data

m4ksio committed:
Polish test, remove unnecessary Cosmos SDK layers from code which does off-chain bundles injection

m4ksio committed:
Remove unused MEV commands and protobuf target.

Clean out some commented out code

m4ksio committed:
Pull-based MEV

m4ksio committed:
Add docker compose integration, and integration test

tychoish commented on app/app.go:
```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
+		if selectedTxsTotalSize+int64(len(tx)) <= maxTxBytes {
+			selectedTxs = append(selectedTxs, tx)
+			selectedTxsTotalSize += int64(len(tx))
+		}
```
if selectedTxTotalSize + <txsize> >= maxTxBytes then we should break from the loop.

m4ksio replied:
Good idea, added

tychoish commented on app/app.go:
```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
+		if selectedTxsTotalSize+bundleSize <= maxTxBytes {
+			// Add all transactions from this bundle
+			for _, tx := range bundle.Transactions {
+				//txBytes := []byte(txStr)
```
why comment

m4ksio replied:
Removed

tychoish commented on app/app.go:
```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
+	bundles := app.MevKeeper.PendingBundles(ctx.BlockHeight())
+
+	if len(bundles) > 0 {
+		// TODO remove this, just for debug
```
before committing?

tychoish commented on app/app.go:
```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
+	}
+
+	// Convert selectedTxs to TxRecords
+	txRecords := make([]*abci.TxRecord, len(selectedTxs))
+	for i, tx := range selectedTxs {
```
why do we build this intermediate slice (`selectedTxs`) when we could just build up txRecords incrementally as we go?

m4ksio replied:
Yep, good idea

tychoish commented on app/app.go:
```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
+		}
+	}
+
+	//TODO purge used bundles, purge expired bundles
```
is this blocking merging?

silostaking replied:
fwiw i think its okay cause technically it only stores for blockheight (and i think this is done at the cleanup), but it would be fine to implement here

m4ksio replied:
We don't really do cleanup yet. But I'm wondering where exactly it should be in the codepath. Can block preparation be retried? Ideal place would be to purge all heights below a finalized block, regardless of who produced it, but that's probably for another PR

tychoish commented on app/app.go:
```diff
@@ -1849,6 +1945,22 @@ func (app *App) RegisterTendermintService(clientCtx client.Context) {
 		}
 	}
+
+	//if app.mevConfig.ListenAddr != "" {
+	//	err := mevbase.StartServer(app.Logger(), app.mevConfig, app.MevKeeper, ctxProvider)
+	//	if err != nil {
+	//		panic(err)
+	//	}
+	//}
```
uncomment or delete?

m4ksio replied:
done

tychoish commented on app/app.go:
```diff
@@ -1849,6 +1945,22 @@ func (app *App) RegisterTendermintService(clientCtx client.Context) {
+	if app.mevConfig.ServerAddr != "" {
+		_, err := mevbase.NewPoller(app.Logger(), app.mevConfig, app.MevKeeper, func() int64 { return app.LastBlockHeight() })
+		if err != nil {
+			panic(err)
```
panic seems aggressive here. 

does this get called for every block height? is the poller something that starts in the background? is there anything that makes sure that it shuts down?

m4ksio replied:
`panic` is what every other module does here in case of error.
`Poller` is created and started once here. It polls in the background, currently simply every hardcoded 200ms. 
No shutdown sequence, as I couldn't find a shutdown sequence in the codebase.

tychoish commented on cmd/seid/cmd/mev.go:
```diff
@@ -0,0 +1,28 @@
+	cmd.PersistentFlags().StringP(tmcli.OutputFlag, "o", "text", "Output format (text|json)")
+
+	cmd.AddCommand(cli.CmdPendingBundles())
+	//cmd.AddCommand(cli.CmdSubmitBundles())
```
deal with this.

m4ksio replied:
dealt with

tychoish commented on x/mev/module.go:
```diff
@@ -0,0 +1,203 @@
+// ValidateGenesisStream performs genesis state validation for the mev module in a streaming fashion.
+func (am AppModuleBasic) ValidateGenesisStream(cdc codec.JSONCodec, config client.TxEncodingConfig, genesisCh <-chan json.RawMessage) error {
+	for genesis := range genesisCh {
+		err := am.ValidateGenesis(cdc, config, genesis)
+		if err != nil {
```
I'd put this on one line to avoid putting pressure on the GC

m4ksio replied:
Not sure is GC is even involved in a local variable like this, but I'm open to suggestions

tychoish commented on x/mev/module.go:
```diff
@@ -0,0 +1,203 @@
+
+// ValidateGenesisStream performs genesis state validation for the mev module in a streaming fashion.
+func (am AppModuleBasic) ValidateGenesisStream(cdc codec.JSONCodec, config client.TxEncodingConfig, genesisCh <-chan json.RawMessage) error {
+	for genesis := range genesisCh {
```
should we attempt to do this in parallel/fan out? is the channel always closed? is there actually any mev gensis state?

m4ksio replied:
No MEV genesis. But no other module parallelize this, apart from EVM which only does two separate routines for unmarshalilng and validation. I doubt we need extra complexity here, not yet.

tychoish commented on x/mev/module.go:
```diff
@@ -0,0 +1,203 @@
+		//	res, err := msgServer.AssociateContractAddress(sdk.WrapSDKContext(ctx), msg)
+		//	return sdk.WrapServiceResult(ctx, res, err)
+		default:
+			errMsg := fmt.Sprintf("unrecognized %s message type MEV MEV MEV: %T", types.ModuleName, msg)
+			return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, errMsg)
```
combine into one statement.

tychoish commented on x/mev/module.go:
```diff
@@ -0,0 +1,203 @@
+	ch := make(chan json.RawMessage)
+	go func() {
+		ch <- am.ExportGenesis(ctx, cdc)
+		close(ch)
```
defer this?

tychoish commented on x/mev/module.go:
```diff
@@ -0,0 +1,203 @@
+func (am AppModule) ExportGenesisStream(ctx sdk.Context, cdc codec.JSONCodec) <-chan json.RawMessage {
+	ch := make(chan json.RawMessage)
+	go func() {
+		ch <- am.ExportGenesis(ctx, cdc)
```
send with a select against the ctx to avoid a deadlock.

m4ksio replied:
As above, boilerplate and how most modules do this. But I wonder how would this possibly deadlock? Since it runs in a goroutine it should wait until somebody reads it

tychoish commented on x/mev/module.go:
```diff
@@ -0,0 +1,203 @@
+// ExportGenesisStream returns the mev module's exported genesis state as raw JSON bytes in a streaming fashion.
+func (am AppModule) ExportGenesisStream(ctx sdk.Context, cdc codec.JSONCodec) <-chan json.RawMessage {
+	ch := make(chan json.RawMessage)
+	go func() {
```
the mustMarshal (which presumably panics if there's an error.) inside of a go routine seems seems like a weird model. is this code actually ever called or is it boilerplate?

m4ksio replied:
Its boilerplate, and that's what most modules are doing

tychoish commented on x/mev/module.go:
```diff
@@ -0,0 +1,203 @@
+func (am AppModule) InitGenesis(ctx sdk.Context, cdc codec.JSONCodec, data json.RawMessage) []abci.ValidatorUpdate {
+	var genesisState types.GenesisState
+	cdc.MustUnmarshalJSON(data, &genesisState)
+	return []abci.ValidatorUpdate{}
```
why return an empty slice rather than nil? 

does this actually have any side effects?

m4ksio replied:
Interestingly, other modules returns empty slice rather then `nil`. No idea why.
In theory, I think that if you had a type alias and declared method it would make a difference if you case `nil` slice and empty one, but I don't see anything like this going on here.

tychoish commented on mev/poll.go:
```diff
@@ -0,0 +1,62 @@
+	logger            log.Logger
+}
+
+func (p *Poller) run() {
```
does anything shut this down?

silostaking replied:
best MR comment of 2025

tychoish replied:
😂 😞

tychoish commented on mev/poll.go:
```diff
@@ -0,0 +1,62 @@
+	logger.Info("Starting bundle provider poller")
+
+	// TODO secure grpc connection
+	grpcConn, err := grpc.DialContext(context.Background(), config.ServerAddr, grpc.WithInsecure())
```
can we pass a real context here so things can exit reasonably

m4ksio replied:
I agree that some shutdown mechanism should be added, but I cannot find one actually used by the application. `sdk.Context` contains `context.Context` which can be used, but I don't think this ever gets cancelled anywhere. Similarly, `App` has `HandleClose` method, but it doesn't seem to be called during shutdown. 
Anyway, added `context.Context` as a parameter and shutdown codepath

tychoish commented on mev/poll.go:
```diff
@@ -0,0 +1,62 @@
+
+	go func() {
+		for range ticker.C {
+			p.run()
```
should get some kind of application context.

I'm not sure if there's one of these pollers per height or not.

tychoish commented on mev/poll.go:
```diff
@@ -0,0 +1,62 @@
+		logger:            logger,
+	}
+
+	ticker := time.NewTicker(200 * time.Millisecond)
```
the interval probably should be a constant or configurable.

silostaking replied:
+1

silostaking replied:
also i think we talked about moving this to GRPC streaming before submitting?

if its easy feel free to rip it but if not its an okay placeholder

m4ksio replied:
Extracted to a const. This value is correlated with block production rate, so I don't think it should be configurable.
I would leave switch to streaming as an optimization in another PR

tychoish commented on app/app.go:
```diff
@@ -1849,6 +1945,22 @@ func (app *App) RegisterTendermintService(clientCtx client.Context) {
+
+	fmt.Printf("MEV RPC Poller address %s\n", app.mevConfig.ServerAddr)
+	if app.mevConfig.ServerAddr != "" {
+		_, err := mevbase.NewPoller(app.Logger(), app.mevConfig, app.MevKeeper, func() int64 { return app.LastBlockHeight() })
```
you could just pass `app.LastBlockHeight` rather than wrapping it in a function

silostaking replied:
Okay, so I think i'm fine with having one ServerAddr for the moment, but we (ideally) need this to be able to take in a list of serveraddresses (all the bundle servers) that way ultimately the community can decide as many as they want to include.

if this is an easy edit for now make it a list. if you want to make a hyperparameter of maximum number of servers that would be fine (and set it to 1 for now so no one goes funky on us)

m4ksio replied:
> you could just pass `app.LastBlockHeight` rather than wrapping it in a function

Yes, great spot, such a silly mistake.

> but we (ideally) need this to be able to take in a list of serveraddresses (all the bundle servers) that way ultimately the community can decide as many as they want to include.

I don't think that a good idea, as it opens an array of problems - ie how to handle conflicting bundles, or which are more important. If a validator wants to handle more then one bundle-server I think it should rather have, say, aggregating bundle-server which does that and provide single stream for a validator.
Unless I misunderstood the comment and we want to have a fallback option, so you can specify multiple servers in case one goes down.

tychoish reviewed (commented):
mostly nits. there's still a few things with the poller and a fair amount of commented out code that looks like it needs to be poked.

silostaking commented on app/app.go:
```diff
@@ -543,6 +551,13 @@ func New(
+	app.MevKeeper = mevbase.NewKeeper(
+		appCodec,
+		keys[mevtypes.StoreKey],
+	)
```
do you need this store anymore? I think we removed all the storeKey/KSStore stuff.

m4ksio replied:
It's handy. For example `PrepareProposalHandler` is method of `app` type so it can easily access `MevKeeper`.

silostaking commented on app/app.go:
```diff
@@ -809,6 +830,7 @@ func New(
 		wasm.ModuleName,
 		tokenfactorytypes.ModuleName,
 		acltypes.ModuleName,
+		mevtypes.ModuleName, // Add here
```
nit: drop the add here on both these now, might as well get this ready to go public

silostaking commented on app/app.go:
```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
+	}
+
+	// Next, add bundle transactions with the highest priority
+	for _, bundle := range bundles {
```
i think this is fine for now; but it doesn't handle:
* What if there are multiple bundles for that blockheight (or is the idea that the `bundles` list is sorted by priority? right now its just time of submission
* There is no cap on the number of bundles submitted.

My ultimate preference here would be (for now):
* Add a variable config setting in the mevkeeper for `max_number_of_bundles_per_block`. Make it like 2 initiallty - so we can test and such. 
* Remove this docstring of `highest priority`, you can just say `first submitted` or `first in the list`.

m4ksio replied:
You are right, in this PR at this stage we order them essentially by submission time. There are two problems
1. we don't have a simulation part so we can't know the tip. 
2. There is also no cap on number of bundles per se, but we limit total byte size of a block by `req.MaxTxBytes` which is arguably better as it better utilize the space, prioritizing bundles over rest of txs

silostaking commented on app/app.go:
```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
+	// This could check for specific message types that should always be included
+	// like governance votes, IBC packets, etc.
+	return false
+}
```
Just flagging we should talk to the Sei team about this - effectively ask what transactions do they want to be omegaadmin and executed first. If you look in the `priority.go` file there are 2 that they elevate - the oracle feeds, and the evm associations - so they might belong in here

silostaking replied:
Maybe update the docstring to something that suggests this

m4ksio replied:
Good question, there is also `PartitionPrioritizedTxs` at **app/app.go:1502** which does similar prioritization, although it seems like it's used to run some transactions first, but respect their position in a block. 
Certainly we need more input on how to implement `isSystem`

silostaking commented on mev/cli/common.go:
```diff
@@ -0,0 +1,26 @@
+//	mevRpcClient := mev.NewMevRpcServiceClient(grpcConn)
+//
+//	return mevRpcClient, clientCtx, nil
+//}
```
do we need this file or can we purge?

m4ksio replied:
purged

silostaking commented on mev/cli/pending_bundles.go:
```diff
@@ -0,0 +1,31 @@
+			//	return err
+			//}
+			//
+			//return clientCtx.PrintProto(&mev.QueryPendingBundlesResponse{Bundles: res.Bundles})
```
Maybe put a todo on these comments; and later on we should implement this. Naively it can just return all bundles, but then later we can add arguments (bundles before height, after height, etc.)

m4ksio replied:
Purged for now, we can restore this once we have an actual use

silostaking commented on mev/cli/submit_bundle.go:
```diff
@@ -0,0 +1,39 @@
+//	}
+//
+//	return cmd
+//}
```
im fine droping this file for now.

We have it saved in git history and let's not confuse folks on the PR

silostaking commented on mev/keeper.go:
```diff
@@ -0,0 +1,97 @@
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	_ sdk.StoreKey, // keep parameter to maintain compatibility but don't use it
```
yeah can drop now

silostaking commented on mev/server.go:
```diff
@@ -0,0 +1,63 @@
+//	}
+//
+//	return nil
+//}
```
can we drop this?

silostaking commented on mev/types.go:
```diff
@@ -0,0 +1,11 @@
+//		BlockNum:  msg.BlockNum,
+//		Timestamp: msg.Timestamp,
+//	}
+//}
```
sam

silostaking commented on proto/README.md:
````diff
@@ -1,5 +1,17 @@
+git ls-files -m '*.pb.go' '*.pb.gw.go' | xargs git checkout -- && git checkout -- go.mod && git checkout -- go.sum && gvm use go1.23 && sed -i '' 's/^go 1\.21$/go 1.23/' go.mod && ignite generate proto-go -y && gvm use go1.21 && git checkout -- go.mod && git checkout -- go.sum && make install && go test ./...
+```
+
+## Remaining notes:
````
so fwiw if you have a better hack around this (or you think im wrong) feel free to delete this. i did this because protos work in this repo was hell and wanted to help future people but maybe im bad at coding

silostaking replied:
fwiw this was the recommendation by their eng team so probably is fine

m4ksio replied:
Frankly, I was able to generate the same protobuf files without switching go versions, just by using `buf` in an properly ancient version. But if that's recommended by SEI team, I think we should keep it.

m4ksio committed:
Review comments
**Remove MEV-related code and unused functionality**

Eliminated MEV module components, including commands, server logic, and protobuf definitions. This cleanup removes unused MEV polling mechanisms and simplifies related areas in the application logic.
